---
title: "Project_lab2"
author: "Marina Panteli"
date: "10/02/2022"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Filtering vcf files to keep only high coverage variants

```{r pressure, echo=FALSE}
library(tibble)
library(VariantAnnotation)
(v <- readVcf("deepvariant_calls_pass.vcf.gz"))
all=geno(v)$DP>10
#View(all)
#typeof(all)
all<-as.data.frame(all)
all1 <- tibble::rownames_to_column(all, "chr_var")
#View(all1)
all2<-all1[all1$default=="TRUE",]
#View(all2)
#table(all2$default)
```
## Filtering vcf file to keep only exonic regions
```{r pressure, echo=FALSE}

library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(stringr)

txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
exons<-exons(txdb)
exons<-as.data.frame(exons)

#get df of exon ranges

dim_length <- dim(all2)[1]
var_location <- rep(NA, dim_length)

a<-all2[,][1]
for (i in 1:dim(a)[1]) {
  k<-a[[1]][i]
  res <- str_match(k, ":\\s*(.*?)\\s*_")
  var_location_now<-res[,2]
  var_location[i]<-as.numeric(var_location_now)
}

is_var_in_exon<-rep(NA, dim_length)

for (j in 1:dim_length) {
  
  is_var_in_exon[j] <- (any((var_location[j]>exons$start) & (var_location[j]<exons$end)))
}


df_is_var_in_exon <- as.data.frame(is_var_in_exon)

df_new <- cbind(all2, df_is_var_in_exon)
df_filtered_variants<-df_new[df_new$is_var_in_exon != "FALSE", ]
df_filtered_vars<- df_filtered_variants[,1]

```

## Find original amino acid sequences

```{r pressure, echo=FALSE}
library(BSgenome.Hsapiens.UCSC.hg38)
library(Biostrings)
x <- import("wtc11_with_cds_refined.gtf")
x <- x[x$type=="CDS"]

gs <- getSeq(Hsapiens, x)


trscr_id<-as.data.frame(x)
trscr_id<-trscr_id[,11]

l<-1

m<-1

trscr_idxes <- vector()

while (m<length(trscr_id)) {
  
  if (trscr_id[l+1]==trscr_id[l]) {
    
    l<-l+1
  }else {
    
    trscr_idxes <- c(trscr_idxes,l)
    l<-l+1
    
  }
 m<-m+1   
  
}


n<- 1
p<-1
d<-vector()

while (p<=length(trscr_idxes)) {
  
d <- c(d,DNAString(paste0(as.character(gs[n:trscr_idxes[p]]), collapse="")) )

n<-trscr_idxes[p]+1

p<-p+1

}


original_seq<-vector()

for (q in 1:length(d)) {
  
  original_seq <- c(original_seq,as.character(translate(DNAString(as.character(d[[q]])))))
  
}


#translate(DNAString("ATG"))


#getSeq(Hsapiens, names="chr1", start=15820, width=10)


```

## Find changed nucleotide sequences (SNPs)
```{r pressure, echo=FALSE}

var_length <- vector()
orig_length <- vector()

for (r in 1:length(df_filtered_vars)) {
 
  a1 <- df_filtered_vars[r]

  orig_length[r] <-gregexpr(pattern ='/',a1)[[1]][1]-gregexpr(pattern ='_',a1)[[1]][1]-1

  var_length[r] <- str_length(a1)-gregexpr(pattern ='/',a1)[[1]][1] 

}

SNP_detection <- vector()

for (r in 1:length(df_filtered_vars)) {
 
  if (orig_length[r]==var_length[r]) {
    
    SNP_detection [r]<- TRUE
    
  }else{
    
    SNP_detection[r]<- FALSE
  }

}

table(SNP_detection)/sum(table(SNP_detection)) #percentage of SNPs

gs_new <- gs

#
s <-1
#
#for (s in 1:length(SNP_detection)) {
  
  if (SNP_detection[s] == TRUE) {
    
    for (t in 1:length(gs)) {
      
      if ((as.numeric (str_match(df_filtered_vars[s], ":\\s*(.*?)\\s*_")[2])<=gs[t]@ranges@start+gs[t]@ranges@width-1) && (as.numeric (str_match(df_filtered_vars[s], ":\\s*(.*?)\\s*_")[2])>=gs[t]@ranges@start)) {  
        
        
        gs_new[[t]][as.numeric (str_match(df_filtered_vars[s], ":\\s*(.*?)\\s*_")[2])-gs_new[t]@ranges@start+1] <- substr(sub("^[^/]*", "", df_filtered_vars[s]),2,2)       ####
        
        t
        s
      }
      
    }
    
  }
  
#}


```
## Find changed amino acid sequence


```{r pressure, echo=FALSE}
library(BSgenome.Hsapiens.UCSC.hg38)
library(Biostrings)
x <- import("wtc11_with_cds_refined.gtf")
x <- x[x$type=="CDS"]




trscr_id<-as.data.frame(x)
trscr_id<-trscr_id[,11]

l<-1

m<-1

trscr_idxes <- vector()

while (m<length(trscr_id)) {
  
  if (trscr_id[l+1]==trscr_id[l]) {
    
    l<-l+1
  }else {
    
    trscr_idxes <- c(trscr_idxes,l)
    l<-l+1
    
  }
 m<-m+1   
  
}


n<- 1
p<-1
d<-vector()

while (p<=length(trscr_idxes)) {
  
d <- c(d,DNAString(paste0(as.character(gs_new[n:trscr_idxes[p]]), collapse="")) )

n<-trscr_idxes[p]+1

p<-p+1

}


new_seq<-vector()

for (q in 1:length(d)) {
  
  new_seq <- c(new_seq,as.character(translate(DNAString(as.character(d[[q]])))))
  
}
```

