get_read_ranges <- function(ga) {
  rngs <- cigarRangesAlongReferenceSpace(cigar(ga), 
                                         pos = start(ga),
                                         drop.empty.ranges = FALSE,
                                         N.regions.removed = FALSE,
                                         with.ops = TRUE,
                                         reduce.ranges = FALSE)
  names(rngs) <- names(ga)
  # remove 'N' ranges
  rngs <- lapply(rngs, function(u) {
    nm <- names(u)
    names(u) <- NULL
    u[nm != "N"]
  })
  IRangesList(rngs)
}


extract_exons<-function(x){
  x_ex <- x[x$type=="exon" & seqnames(x) != "chrM"]
  g <- grepl(gene, x_ex$transcript_id, 
             fixed = TRUE)
  x_ex <- x_ex[g]
  seqlevels(x_ex) <- seqlevels(x_ex)[-25]
  x_ex
}

narrowal<-function(st,rds,ga,chromosome){
  
  na<-vector()
  df<-vector()  
  
  for (i in st) {
    
    (gr <- GRanges(chromosome,
                   IRanges(start = i,
                           width = 2),
                   strand = "+"))
    
    #read in BAM alignments only for this region
    ga_k <- ga[names(ga) %in% rds] 
    na <- narrowAlignments(ga_k, gr)
    emd <- elementMetadata(na)
    df <- c(df, as.data.frame(emd)) 
    
  }
  
  df
}
seqs_with_var<-function(to_insert,ref_seq,mtt,x_ex,k,new_seqs){
  seqs <- list()
  all_seqnames <- vector()
  for(i in 1:length(to_insert)) { # loop through alleles
    seqs[[i]] <- ref_seq
    for(j in length(mtt):1) {     # loop through variations in 3'->5' direction
      subseq(seqs[[i]], start=start(mtt)[j], 
             end=start(mtt)[j]+1) <- to_insert[[i]][j]
      # remove '-'
      seqs[[i]] <- DNAStringSet(gsub("-","",seqs[[i]]))
    }
    names(seqs[[i]]) <- paste0(names(seqs[[i]]), ".", letters[i])
    all_seqnames<-c(all_seqnames,names(seqs[[i]]))
  }
  seqs <- unlist(DNAStringSetList(seqs))
  
  if (as.logical(strand(x_ex[x_ex$transcript_id==k]) == "-")[1]) {
    seqs<-reverseComplement(seqs)
  } 
  
  new_seqs<-c(new_seqs,seqs)
  new_seqs
}

get_alleles<-function(df){
  z<-matrix(unlist(df), ncol=length(df))
  alleles <- table(apply(z, 1, paste, collapse="__")) # The different combinations that are generated by the different reads
  alleles <- sort(alleles, decreasing = TRUE)
  alleles <- alleles[seq_len(min(2,length(alleles)))]
  (to_insert <- strsplit(names(alleles), "__", fixed = TRUE))
  to_insert
}