get_read_ranges <- function(ga) {
  rngs <- cigarRangesAlongReferenceSpace(cigar(ga), 
                                         pos = start(ga),
                                         drop.empty.ranges = FALSE,
                                         N.regions.removed = FALSE,
                                         with.ops = TRUE,
                                         reduce.ranges = FALSE)
  names(rngs) <- names(ga)
  # remove 'N' ranges
  rngs <- lapply(rngs, function(u) {
    nm <- names(u)
    names(u) <- NULL
    u[nm != "N"]
  })
  IRangesList(rngs)
}


extract_exons <- function(x, gene) {
  x_ex <- x[x$type=="exon" & seqnames(x) != "chrM"]
  g <- grepl(gene, x_ex$transcript_id, 
             fixed = TRUE)
  x_ex <- x_ex[g]
  w <- which(seqlevels(x_ex)=="chrM")
  seqlevels(x_ex) <- seqlevels(x_ex)[-w]
  x_ex
}

narrowal <- function(starts, ga){
  
  lapply(starts, function(u) {
    ir <- IRanges(start = u, width = 2)
    gr <- GRanges(seqnames(ga)[1],
                  ir, strand = "+")
    na <- narrowAlignments(ga, gr)
    as.character(mcols(na)$seq)
  })
  
}


seqs_with_var <- function(to_insert,ref_seq,mtt,x_ex,tr){
  seqs <- list()
  all_seqnames <- vector()
  for(i in 1:length(to_insert)) { # loop through alleles
    seqs[[i]] <- ref_seq
    for(j in length(mtt):1) {     # loop through variations in 3'->5' direction
      subseq(seqs[[i]], start=start(mtt)[j], 
             end=start(mtt)[j]+1) <- to_insert[[i]][j]
      # remove '-'
      seqs[[i]] <- DNAStringSet(gsub("-","",seqs[[i]]))
    }
    names(seqs[[i]]) <- paste0(names(seqs[[i]]), ".", letters[i])
    all_seqnames<-c(all_seqnames,names(seqs[[i]]))
  }
  seqs <- unlist(DNAStringSetList(seqs))
  
  if (as.character(strand(x_ex)[1]) == "-")
    seqs <- reverseComplement(seqs)
  
  seqs
}

get_alleles <- function(df){
  z <- matrix(unlist(df), ncol=length(df))
  alleles <- table(apply(z, 1, paste, collapse="__")) # The different combinations that are generated by the different reads
  alleles <- sort(alleles, decreasing = TRUE)
  alleles <- alleles[seq_len(min(2,length(alleles)))]
  strsplit(names(alleles), "__", fixed = TRUE)
}

generate_variant_transcripts <- function(v, x, 
                                         gene = "PB.749.", 
                                         bam_file = "aln_s.bam", verbose = FALSE) {
  
  if(verbose)
    message(paste0("working on '", gene, "'"))
  
  # extract all exons for given gene  
  x_ex <- extract_exons(x, gene) ##
  
  # find overlaps b/w variation and this gene
  fop <- findOverlapPairs(rowRanges(v), x_ex) 
  
  #check that there IS an overlap
  if (!isEmpty(pintersect(fop))) {
  
    # read alignments for this region
    gr <- range(x_ex)
    sbp <- ScanBamParam(which = gr, what = scanBamWhat())
    bamWhat(sbp) <- "seq"
    ga <- readGAlignments(bam_file, 
                          param = sbp, use.names = TRUE)
    
    # full set of variants for this region
    variation_regs <- unique(first(fop))
    
    # ranges for each read
    rngs <- get_read_ranges(ga) ##
    
    # split into transcripts
    x_exs <- split(x_ex, x_ex$transcript_id)
    
    # intersect all reads w/ all transcripts, form matrix
    m <- matrix(0,
                ncol = length(variation_regs),
                nrow = length(rngs),
                dimnames = list(names(rngs), names(variation_regs))
    )
    rngs_u <- unlist(rngs)
    fo <- findOverlaps(rngs_u, ranges(variation_regs))
    
    # which variations in a certain alignment read, and 
    # below notate with 0/1 if absent/present
    pop <- split(subjectHits(fo),
                 names(rngs_u)[queryHits(fo)]) 
    for (i in 1:length(pop)) {
      m[names(pop)[i], unique(pop[[i]])] <- 1
    }
    
    transcripts <- unique(second(fop)$transcript_id)
    new_seqs <- vector("list", length(transcripts))
    
    for (i in seq_len(length(transcripts))) {
      tr <- transcripts[i]
      if(verbose)
        message(tr)
      
      f <- first(fop)[second(fop)$transcript_id == tr]
      nm <- names(f)
      
      # only take reads that overlap all variation pos. of this transcript
      w <- rowSums(m[, nm, drop = FALSE]) == length(nm)
      rds <- rownames(m)[w]
      
      # extract transcript seq from genome + 
      # positions where we must insert variation (transcript coord.)
      dss_ref <- getSeq(Hsapiens, x_ex[x_ex$transcript_id == tr])
      ref_seq <- DNAStringSet(paste(as.character(dss_ref),
                                    collapse = ""))
      names(ref_seq) <- tr
      mtt <- mapToTranscripts(variation_regs[nm], x_exs[tr])
      st <- start(variation_regs[nm])
      
      df <- narrowal(st, ga[names(ga) %in% rds])
      to_insert <- get_alleles(df)
      new_seqs[[i]] <- seqs_with_var(to_insert, ref_seq, 
                                     mtt, x_ex, tr)
    }
    new_seqs
  }
  


  # check for transcripts with NO overlap
  if(!isEmpty(setdiff(names(table(x_ex$transcript_id)),transcripts))){

    seqs_old <- vector("list", length(setdiff(names(table(x_ex$transcript_id)),transcripts)))
    transcripts_old <- setdiff(names(table(x_ex$transcript_id)),transcripts)

    for (j in seq_len(length(transcripts_old))) {
      tr_old <- transcripts_old[j]
      if(verbose)
        message(tr_old)

      dss_ref_old <- getSeq(Hsapiens, x_ex[x_ex$transcript_id == tr_old])
      ref_seq_old <- DNAStringSet(paste(as.character(dss_ref_old),
                                        collapse = ""))
      names(ref_seq_old) <- tr_old

      if(as.character(strand(x_ex[x_ex$transcript_id == tr_old])[1])=='-'){
        ref_seq_old<-reverseComplement(ref_seq_old)
      }

      seqs_old[[j]] <- ref_seq_old
    }
    seqs_old
  }

if(length(seqs_old)!=0 && length(new_seqs)!=0){
  new_seqs<- do.call(c,list(new_seqs,seqs_old))
} else if (length(seqs_old)!=0 && length(new_seqs)==0){
  new_seqs <- seqs_old
}else if (length(seqs_old)==0 && length(new_seqs)!=0){
  new_seqs <- new_seqs
}
  new_seqs
}
